app/webdav/webdav.go

```diff
@@ -0,0 +1,407 @@
+	br blob.Ref
+
+	mu  sync.Mutex
+	fi  os.FileInfo
+	sub map[string]*webdavFsNode
+
+	// cache invalidation data
+	static        bool
+	lastRefreshed time.Time
+}
+
+var refreshInterval = 1 * time.Minute
+
+func needsRefresh(n *webdavFsNode) bool {
+	return !n.static && time.Now().After(n.lastRefreshed.Add(refreshInterval))
+}
+
+func (fs *webdavFs) refreshRoot(ctx context.Context, n *webdavFsNode) error {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	if !needsRefresh(n) {
+		return nil
+	}
+
+	des, err := fs.client.Describe(ctx, &search.DescribeRequest{BlobRef: n.br, Depth: 3})
+	if err != nil {
+		return fmt.Errorf("unable to describe blob ref %s: %w", n.br, err)
+	}
+	db := des.Meta.Get(n.br)
+
+	if db.CamliType != schema.TypePermanode {
+		return fmt.Errorf("root %s should be a permanode", n.br)
+	}
+
+	sub := make(map[string]*webdavFsNode, 0)
+	for k := range db.Permanode.Attr {
+		if !strings.HasPrefix(k, nodeattr.CamliPathColon) {
+			continue
+		}
+		cb := blob.ParseOrZero(db.Permanode.Attr.Get(k))
+		if !cb.Valid() {
+			continue
+		}
+		name := strings.TrimPrefix(k, nodeattr.CamliPathColon)
+		dbm := des.Meta.Get(cb)
+		sub[name] = &webdavFsNode{br: dbm.BlobRef}
+	}
+	for k, v := range sub {
+		if c, ok := n.sub[k]; ok {
+			if c.br == v.br {
+				continue
+			}
+		}
+		n.sub[k] = v
+	}
+	for k, v := range n.sub {
+		if _, ok := sub[k]; !ok {
+			sub[k] = v
+		}
+	}
+
+	n.static = false
+	n.sub = sub
+	n.fi = webdavFsFileInfo{
+		isDir:   true,
+		name:    "/",
+		mode:    0400,
+		size:    int64(len(n.sub)),
+		modTime: time.Now(),
+	}
+	n.lastRefreshed = time.Now()
+	return nil
+}
+
+func (fs *webdavFs) refresh(ctx context.Context, n *webdavFsNode) error {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	if !needsRefresh(n) {
+		return nil
+	}
+
+	des, err := fs.client.Describe(ctx, &search.DescribeRequest{BlobRef: n.br, Depth: 3})
+	if err != nil {
+		return fmt.Errorf("unable to describe blob ref %s: %w", n.br, err)
+	}
+	db := des.Meta.Get(n.br)
+
+	switch db.CamliType {
+	case schema.TypePermanode:
+		// TODO
+		return fmt.Errorf("unable to refresh permanodes %s: %w", n.br, err)
+	case schema.TypeFile:
+		n.static = true
+		n.fi = webdavFsFileInfo{
+			isDir:   false,
+			name:    db.File.FileName,
+			size:    db.File.Size,
+			mode:    0400,
+			modTime: modtimeFromFileInfo(db.File),
+		}
+	case schema.TypeDirectory:
+		n.static = true
+		n.fi = webdavFsFileInfo{
+			isDir:   true,
+			name:    db.Dir.FileName,
+			size:    db.Dir.Size,
+			mode:    0400,
+			modTime: modtimeFromFileInfo(db.Dir),
+		}
+		n.sub = make(map[string]*webdavFsNode)
+		for _, m := range db.DirMembers() {
+			dmc := des.Meta.Get(m.BlobRef)
+			var fi *camtypes.FileInfo
+			if dmc.File != nil {
+				fi = dmc.File
+			} else if dmc.Dir != nil {
+				fi = dmc.Dir
+			}
+			if fi == nil {
+				continue
+			}
+			n.sub[fi.FileName] = &webdavFsNode{br: m.BlobRef}
+		}
+	}
+	n.lastRefreshed = time.Now()
+	return nil
+}
+
+type webdavFsFile struct {
```

Comment:
maybe get "RO" (for Read Only) into this type name? And then some docs on the type.

Reply:
:heavy_check_mark: